assemblers/asl-1.41r6/p2bin.1
.TH p2bin 1

.SH NAME
.B p2bin
\- convert code files into hex files

.SH SYNTAX
.B p2bin
[ option(s) ] <name(s)> [ further options/names ]

.SH DESCRIPTION

P2BIN is a tool to convert the contents of one or several code files
generated by AS into binary files.  A binary file is a 1:1 memory image
of the processor's memory and is especially suited for EPROM programmers
and emulators.

Arguments to P2BIN may be either command line parameters or file name
specifications.  Any argument that starts with the charactes +, - or
/ is regarded as a comand line parameter (which may take an
additional command line argument); any other argument is regarded as
a file name.  Generally, P2BIN needs at least two file names: An
input code file and the name of the binary output file.  If multiple
file names are given, P2BIN will always take the last name as the
output file's name.  If an input file name does not have an
extension, the extension '.p' is added automatically.  Similarly, the
extension '.bin' is added automatically to the target file's name.
A special case occurs when only one file name is given: P2BIN will
then take its name as the source (possibly extended with '.p'), and
the same name as target (with '.bin' as additional or replaced
extension).

.SH COMMAND-LINE PARAMETERS

If a command-line parameter starts with a slash(/) or minus sign(-),
it turns an option on; if a command-line parameter starts with a plus
sign(+), it turns a specific option off.  Numeric arguments to
parameters can be either written in decimal or hexadecimal notation.
For hexadecimal notation, prefix the number with a dollar($) sign.
In the following list, all options will be shown in the form that is
needed to change the default behaviour, which might be a plus or
minus sign, depening on wether the option is on or off by default.

.B p2bin
accepts the following command-line parameters:
.TP
.B -f <number>[,<further numbers>]
Add <number> to the list of record header IDs that allow a record
from a source file to be written to the target file.  A certain
header ID marks code for a certain target processor family; thus,
this filter allows to distill code for a certain processor out of a
source file that contains code for different processor families.
Negation of this parameter removes certain header IDs from P2BIN's
list.  See the user manual of AS for a list of all possible header ID
values.  If P2BIN's list of header IDs is empty, no filtering will
take place, i.e. all records from a source file will make it into the
target file.
.TP
.B -l <number>
Set the value that should be used to fill memory areas in the binary
image that are unused by the input code file(s).  The default for this
is to fill unused areas with the value 255 (0xff), which is the best choice
for EPROMs as erased EPROM cells carry this value and an intelligent
EPROM burner can skip these cells, speeding up the programming process and
reducing stress for the EPROM.  However, there are specialized EPROMs that
have zeros in their cells after erasure, and you might want to fill unused
areas with a code that executes as a NOP or BREAK statement.
.TP
.B -m <all|even|odd|byte<0|1|2|3>|word<0|1>>
Set the mask of bytes to be filtered out.  If your target processor has
a 16- or 32-bit data path, but your EPROMs are only 8- or 16-bits wide,
the code has to be spread over the EPROMs in an alternating fashion.
This option allows you to do the necesary splitting, however you have
to run P2BIN two or four times with different masks.  The possible arguments
have the following meanings:

.B all
does not do any filtering, i.e. all bytes of your code will show up in the
resulting image.  This is the default.

.B even
or
.B odd
will take only those bytes whose addresses are in the form 2*n or 2*n+1.  They
are useful if you have a 16-bit data path and two 8-bit EPROMs.

.B byte0, byte1, byte2
or
.B byte3
will take only those bytes whose addresses are in the form 4*n ... 4*n+3.
They are useful if you have a 32-bit data path and four 8-bit EPROMs.

.B word0
or
.B word1
will take only those bytes whose addresses are in the form 4*n+0 / 4*n+1
or 4*n+2 / 4*n+3.  They are useful if you have a 32-bit data path and two
16-bit EPROMs.

When using one of these filters, the resulting images will automatically
become smaller by a factor of 2 or 4.  Beware that this does not influence
address specifications given with the
.B -r
command-line parameter! See also the examples section below for correct
usage.
.TP
.B -r < <start>-<stop> >
Set a certain address range to be filtered out of the input file(s).
Code that lies outside this range does not appear in the output file.
The default for the address filter is the 0-$7fff, which might create
confusion in some cases.  As a special option,
.B <start>
and
.B <stop>
may consist of just a single dollar sign (escape this
in UNIX shells!) to signify the lowest resp. highest address that
occurs in the input file(s).  Using this option will implicitly
enable a second pass over all input files to find the minimum and
maximum values before conversion starts, reducing the speed of P2BIN
slightly.
.TP
.B -s

Tell P2BIN to include a checksum into the image.  A checksum is a byte
value entered into the image's last byte that is the two's complement of
the sum of all previous bytes.  Therefore, the sum of all bytes modulus
256 will become zero.  This option is useful if you want to check the
ROM contents in your program as part of a power-on self-test, but keep
in mind that you must not use the last byte for your own purposes any
more!

.SH PRESETTING PARAMETERS

Parameters need not neccessarily be given in the command line itself.  Before
processing of command line parameters starts, P2BIN will look if the
.B P2BINCMD
environment variable is defined.  If it exists, its contents will be
treated as additional command line paramters whose syntax is absolutely
equal to normal command line parameters.  As exception is made if the
variable's contents start with a '@' sign; in such a case, the string after
the '@' sign is treated as the name of a file that contains the options.
Such a file (also called a 'key file') has the advantage that it allows
the options to be written in different lines, and it does not have a size
limit.  Some operating systems (like MS-DOS) do have a length limit on
command lines and environment variable contents, so the key file may be
your only option if you have a lot of lengthy parameters for P2BIN.

.SH RETURN CODES

.B p2bin
may return with the following codes:
.TP
.B 0
no errors.
.TP
.B 1
incorrect command line parameters.
.TP
.B 2
I/O-error.
.TP
.B 3
An input file had an incorrect format.

.SH EXAMPLES

To convert a file
.B file1.p
fully into its binary representation, use
.PP
.B p2bin -r \e$-\e$ file1
.PP
If you have a processor with a 64 KByte address space and a 16-bit
data path and you want to assure that the memory image always starts
at address 0, regardless of address layout in the code file, use
.PP
.B p2bin -r 0-\e$ffff -m even file1
.B evenfile

.B p2bin -r 0-\e$ffff -m odd file1
.B oddfile
.PP
to get images for two 27256 EPROMs.

.SH TIPS

Calling P2BIN without any arguments will print a short help
listing all command line parameters.

.SH SEE ALSO

asl(1), plist(1), pbind(1), p2hex(1)

.SH HISTORY

P2BIN originally appeared as an AS tool in 1992, written in
Borland-Pascal, and was ported to C and UNIX in 1996.

.SH BUGS

Command line interpreters of some operating systems reserve some
characters for their own use, so it might be necessary to give
command line parameters with certain tricks (e.g., with the help
of escape characters).

P2BIN does not have so far an opportunity to filter records by
target segment.  Instead, records that contain data for any other
segment than CODE are completely ignored.

.SH AUTHOR(S)

Alfred Arnold (a.arnold@kfa-juelich.de)



assemblers/asl-1.41r6/p2bin.c
// AS-Portierung
// Umwandlung von AS-Codefiles in Binaerfiles
#include "stdinc.h"
#include <string.h>
#include <ctype.h>

#include "endian.h"
#include "bpemu.h"
#include "stringutil.h"
#include "hex.h"
#include "nls.h"
#include "chunks.h"
#include "decodecmd.h"
#include "toolutils.h"

#define BinSuffix ".bin"

typedef void (*ProcessProc)(char *FileName, LongWord Offset);

static CMDProcessed ParProcessed;
static Integer z;

static FILE *TargFile;
static String SrcName, TargName;

static LongWord StartAdr, StopAdr, RealFileLen;
static LongWord MaxGran, Dummy;
static bool StartAuto, StopAuto;

static Byte FillVal;
static bool DoCheckSum;

static Byte SizeDiv;
static LongInt ANDMask, ANDEq;

static ChunkList UsedList;

#include "ioerrors.rsc"
#include "tools.rsc"
#include "p2bin.rsc"

#ifdef DEBUG
#   define ChkIO(s) ChkIO_L(s,__LINE__)
static void ChkIO_L(char *s, int line) {
   if (errno != 0) {
      fprintf(stderr, "%s %d\n", s, line);
      exit(3);
   }
}
#endif

static void ParamError(bool InEnv, char *Arg) {
   printf("%s%s\n%s\n", (InEnv) ? ErrMsgInvEnvParam : ErrMsgInvParam, Arg, ErrMsgProgTerm);
   exit(1);
}

#define BufferSize 4096
static Byte Buffer[BufferSize];

static void OpenTarget(void) {
   LongWord Rest, Trans;

   TargFile = fopen(TargName, OPENWRMODE);
   if (TargFile == NULL) ChkIO(TargName);
   RealFileLen = ((StopAdr - StartAdr + 1) * MaxGran) / SizeDiv;

   memset(Buffer, FillVal, BufferSize);

   Rest = RealFileLen;
   while (Rest != 0) {
      Trans = min(Rest, BufferSize);
      if (fwrite(Buffer, 1, Trans, TargFile) != Trans) ChkIO(TargName);
      Rest -= Trans;
   }
}

static void CloseTarget(void) {
   LongWord Sum, Rest, Trans, Real, z;

   if (fclose(TargFile) == EOF) ChkIO(TargName);

   if (DoCheckSum) {
      TargFile = fopen(TargName, OPENUPMODE);
      if (TargFile == NULL) ChkIO(TargName);
      Rest = FileSize(TargFile) - 1;
      Sum = 0;
      while (Rest != 0) {
         Trans = min(Rest, BufferSize);
         Rest -= Trans;
         Real = fread(Buffer, 1, Trans, TargFile);
         if (Real != Trans) ChkIO(TargName);
         for (z = 0; z < Trans; Sum += Buffer[z++]);
      }
      errno = 0;
      printf("%s%s\n", InfoMessChecksum, HexLong(Sum));
      Buffer[0] = 0x100 - (Sum & 0xff);
      fflush(TargFile);
      if (fwrite(Buffer, 1, 1, TargFile) != 1) ChkIO(TargName);
      fflush(TargFile);
      if (fclose(TargFile) == EOF) ChkIO(TargName);
   }

   if (Magic != 0) unlink(TargName);
}

static void ProcessFile(char *FileName, LongWord Offset) {
   FILE *SrcFile;
   Word TestID;
   Byte InpHeader, InpSegment;
   LongWord InpStart, SumLen;
   Word InpLen, TransLen, ResLen;
   bool doit;
   LongWord ErgStart, ErgStop, NextPos;
   Word ErgLen = 0;
   LongInt z;
   Byte Gran;

   SrcFile = fopen(FileName, OPENRDMODE);
   if (SrcFile == NULL) ChkIO(FileName);

   if (!Read2(SrcFile, &TestID)) ChkIO(FileName);
   if (TestID != FileID) FormatError(FileName, FormatInvHeaderMsg);

   errno = 0;
   printf("%s==>>%s", FileName, TargName);
   ChkIO(OutName);

   SumLen = 0;

   do {
      ReadRecordHeader(&InpHeader, &InpSegment, &Gran, FileName, SrcFile);
      if (InpHeader == FileHeaderStartAdr) {
         if (!Read4(SrcFile, &ErgStart)) ChkIO(FileName);
      } else if (InpHeader != FileHeaderEnd) {
         if (!Read4(SrcFile, &InpStart)) ChkIO(FileName);
         if (!Read2(SrcFile, &InpLen)) ChkIO(FileName);

         NextPos = ftell(SrcFile) + InpLen;
         if (NextPos >= FileSize(SrcFile) - 1)
            FormatError(FileName, FormatInvRecordLenMsg);

         doit = (FilterOK(InpHeader) && (InpSegment == SegCode));

         if (doit) {
            InpStart += Offset;
            ErgStart = max(StartAdr, InpStart);
            ErgStop = min(StopAdr, InpStart + (InpLen / Gran) - 1);
            doit = (ErgStop >= ErgStart);
            if (doit) {
               ErgLen = (ErgStop + 1 - ErgStart) * Gran;
               if (AddChunk(&UsedList, ErgStart, ErgStop - ErgStart + 1, true)) {
                  errno = 0;
                  printf(" %s\n", ErrMsgOverlap);
                  ChkIO(OutName);
               }
            }
         }

         if (doit) {
         /* an Anfang interessierender Daten */

            if (fseek(SrcFile, (ErgStart - InpStart) * Gran, SEEK_CUR) == -1) ChkIO(FileName);

         /* in Zieldatei an passende Stelle */

            if (fseek(TargFile, ((ErgStart - StartAdr) * Gran) / SizeDiv, SEEK_SET) == -1) ChkIO(TargName);

         /* umkopieren */

            while (ErgLen > 0) {
               TransLen = min(BufferSize, ErgLen);
               if (fread(Buffer, 1, TransLen, SrcFile) != TransLen) ChkIO(FileName);
               if (SizeDiv == 1) ResLen = TransLen;
               else {
                  ResLen = 0;
                  for (z = 0; z < (LongInt) TransLen; z++)
                     if (((ErgStart * Gran + z) & ANDMask) == ANDEq)
                        Buffer[ResLen++] = Buffer[z];
               }
               if (fwrite(Buffer, 1, ResLen, TargFile) != ResLen) ChkIO(TargName);
               ErgLen -= TransLen;
               ErgStart += TransLen;
               SumLen += ResLen;
            }
         }
         if (fseek(SrcFile, NextPos, SEEK_SET) == -1) ChkIO(FileName);
      }
   }
   while (InpHeader != 0);

   errno = 0;
   printf("  (%d Byte)\n", SumLen);
   ChkIO(OutName);

   if (fclose(SrcFile) == EOF) ChkIO(FileName);
}

static void ProcessGroup(char *GroupName_O, ProcessProc Processor) {
/**   s:SearchRec;**/
   String /**Path,Name,**/ Ext, GroupName;
   LongWord Offset;

   strmaxcpy(GroupName, GroupName_O, 255);
   strmaxcpy(Ext, GroupName, 255);
   if (!RemoveOffset(GroupName, &Offset)) ParamError(false, Ext);
   AddSuffix(GroupName, Suffix);

   Processor(GroupName, Offset);
#if 0
   FSplit(GroupName, Path, Name, Ext);

   FindFirst(GroupName, Archive, s);
   if (DosError)
      WriteLn(ErrMsgNullMaskA, GroupName, ErrMsgNullMaskB);
   else while (!DosError) {
      Processor(Path + s.Name, Offset);
      FindNext(s);
   }
#endif
}

static void MeasureFile(char *FileName, LongWord Offset) {
   FILE *f;
   Byte Header, Gran, Segment;
   Word Length, TestID;
   LongWord Adr, EndAdr, NextPos;

   f = fopen(FileName, OPENRDMODE);
   if (f == NULL) ChkIO(FileName);

   if (!Read2(f, &TestID)) ChkIO(FileName);
   if (TestID != FileMagic) FormatError(FileName, FormatInvHeaderMsg);

   do {
      ReadRecordHeader(&Header, &Segment, &Gran, FileName, f);

      if (Header == FileHeaderStartAdr) {
         if (fseek(f, sizeof(LongWord), SEEK_CUR) == -1) ChkIO(FileName);
      } else if (Header != FileHeaderEnd) {
         if (!Read4(f, &Adr)) ChkIO(FileName);
         if (!Read2(f, &Length)) ChkIO(FileName);
         NextPos = ftell(f) + Length;
         if (NextPos > FileSize(f))
            FormatError(FileName, FormatInvRecordLenMsg);

         if (FilterOK(Header) && (Segment == SegCode)) {
            Adr += Offset;
            EndAdr = Adr + (Length / Gran) - 1;
            if (Gran > MaxGran) MaxGran = Gran;
               if (StartAuto) if (StartAdr > Adr) StartAdr = Adr;
               if (StopAuto) if (EndAdr > StopAdr) StopAdr = EndAdr;
         }

         fseek(f, NextPos, SEEK_SET);
      }
   }
   while (Header != 0);

   if (fclose(f) == EOF) ChkIO(FileName);
}

static CMDResult CMD_AdrRange(bool Negate, char *Arg) {
   char *p, Save;
   bool err;

   if (Arg == NULL); /* satisfy some compilers */

   if (Negate) {
      StartAdr = 0;
      StopAdr = 0x7fff;
      return CMDOK;
   } else {
      p = strchr(Arg, '-');
      if (p == NULL) return CMDErr;

      Save = (*p);
      *p = '\0';
      if ((StartAuto = (strcmp(Arg, "$") == 0))) err = true;
      else StartAdr = ConstLongInt(Arg, &err);
      *p = Save;
      if (!err) return CMDErr;

      if ((StopAuto = (strcmp(p + 1, "$") == 0))) err = true;
      else StopAdr = ConstLongInt(p + 1, &err);
      if (!err) return CMDErr;

      if ((!StartAuto) && (!StopAuto) && (StartAdr > StopAdr)) return CMDErr;

      return CMDArg;
   }
}

static CMDResult CMD_ByteMode(bool Negate, char *Arg) {
#define ByteModeCnt 9
   static char *ByteModeStrings[ByteModeCnt] = { "ALL", "EVEN", "ODD", "BYTE0", "BYTE1", "BYTE2", "BYTE3", "WORD0", "WORD1" };
   static Byte ByteModeDivs[ByteModeCnt] = { 1, 2, 2, 4, 4, 4, 4, 2, 2 };
   static Byte ByteModeMasks[ByteModeCnt] = { 0, 1, 1, 3, 3, 3, 3, 2, 2 };
   static Byte ByteModeEqs[ByteModeCnt] = { 0, 0, 1, 0, 1, 2, 3, 0, 2 };

   Integer z;

   if (Negate); /* satisfy some compilers */

   if (*Arg == '\0') {
      SizeDiv = 1;
      ANDEq = 0;
      ANDMask = 0;
      return CMDOK;
   } else {
      for (z = 0; z < strlen(Arg); z++) Arg[z] = toupper(Arg[z]);
      ANDEq = 0xff;
      for (z = 0; z < ByteModeCnt; z++)
         if (strcmp(Arg, ByteModeStrings[z]) == 0) {
            SizeDiv = ByteModeDivs[z];
            ANDMask = ByteModeMasks[z];
            ANDEq = ByteModeEqs[z];
         }
      if (ANDEq == 0xff) return CMDErr;
      else return CMDArg;
   }
}

static CMDResult CMD_FillVal(bool Negate, char *Arg) {
   bool err;

   if (Negate); /* satisfy some compilers */

   FillVal = ConstLongInt(Arg, &err);
   if (!err) return CMDErr;
   else return CMDArg;
}

static CMDResult CMD_CheckSum(bool Negate, char *Arg) {
   if (Arg == NULL); /* satisfy some compilers */

   DoCheckSum = !Negate;
   return CMDOK;
}

#define P2BINParamCnt 5
static CMDRec P2BINParams[P2BINParamCnt] = {
   { "f", CMD_FilterList },
   { "r", CMD_AdrRange },
   { "s", CMD_CheckSum },
   { "m", CMD_ByteMode },
   { "l", CMD_FillVal }
};

int main(int argc, char **argv) {
   ParamStr = argv;
   ParamCount = argc - 1;
   endian_init();
   stringutil_init();
   bpemu_init();
   hex_init();
   nls_init();
   chunks_init();
   decodecmd_init();
   toolutils_init();

   NLS_Initialize();
   WrCopyRight("P2BIN/C V1.41r5");

   InitChunk(&UsedList);

   if (ParamCount == 0) {
      errno = 0;
      printf("%s%s%s\n", InfoMessHead1, GetEXEName(), InfoMessHead2);
      ChkIO(OutName);
      for (z = 0; z < InfoMessHelpCnt; z++) {
         errno = 0;
         printf("%s\n", InfoMessHelp[z]);
         ChkIO(OutName);
      }
      exit(1);
   }

   StartAdr = 0;
   StopAdr = 0x7fff;
   StartAuto = false;
   StopAuto = false;
   FillVal = 0xff;
   DoCheckSum = false;
   SizeDiv = 1;
   ANDEq = 0;
   ProcessCMD(P2BINParams, P2BINParamCnt, ParProcessed, "P2BINCMD", ParamError);

   if (ProcessedEmpty(ParProcessed)) {
      errno = 0;
      printf("%s\n", ErrMsgTargMissing);
      ChkIO(OutName);
      exit(1);
   }

   z = ParamCount;
   while ((z > 0) && (!ParProcessed[z])) z--;
   strmaxcpy(TargName, ParamStr[z], 255);
   if (!RemoveOffset(TargName, &Dummy)) ParamError(false, ParamStr[z]);
   ParProcessed[z] = false;
   if (ProcessedEmpty(ParProcessed)) {
      strmaxcpy(SrcName, ParamStr[z], 255);
      DelSuffix(TargName);
   }
   AddSuffix(TargName, BinSuffix);

   MaxGran = 1;
   if ((StartAuto) || (StopAuto)) {
      if (StartAuto) StartAdr = 0xffffffffu;
      if (StopAuto) StopAdr = 0;
      if (ProcessedEmpty(ParProcessed)) ProcessGroup(SrcName, MeasureFile);
         else for (z = 1; z <= ParamCount; z++)
            if (ParProcessed[z]) ProcessGroup(ParamStr[z], MeasureFile);
      if (StartAdr > StopAdr) {
         errno = 0;
         printf("%s\n", ErrMsgAutoFailed);
         ChkIO(OutName);
         exit(1);
      }
   }

   OpenTarget();

   if (ProcessedEmpty(ParProcessed)) ProcessGroup(SrcName, ProcessFile);
      else for (z = 1; z <= ParamCount; z++)
         if (ParProcessed[z]) ProcessGroup(ParamStr[z], ProcessFile);

   CloseTarget();
   return 0;
}
